### ISSUE TITLE
Fatal SERVERPOD error in 'serverpod generate'

### ISSUE THREAD

**Issue by Sandyhi5** [serverpod/serverpod#2920]
Message from serverpod asked to post below and create a new Issue.

Problem appeared out of the blue. A seemingly normal code change - generate - migration - run cycle was the plan but during generate got this problem:

⠼ Generating code... (50.8s)
Yikes! It is possible that this error is caused by an internal issue with the Serverpod tooling. We would appreciate if        
you filed an issue over at Github. Please include the stack trace below and describe any steps you did to trigger the error.   
https://github.com/serverpod/serverpod/issues

RangeError (start): Invalid value: Only valid value is 0: 1

#0      RangeError.checkValidRange (dart:core/errors.dart:360:7)
#1      List.sublist (dart:core-patch/growable_array.dart:84:38)
#2      EndpointMethodAnalyzer.parse
(package:serverpod_cli/src/analyzer/dart/endpoint_analyzers/endpoint_method_analyzer.dart:30:39)
#3      EndpointsAnalyzer._parseLibrary (package:serverpod_cli/src/analyzer/dart/endpoints_analyzer.dart:141:47)
#4      EndpointsAnalyzer.analyze (package:serverpod_cli/src/analyzer/dart/endpoints_analyzer.dart:83:27)
<asynchronous suspension>
#5      performGenerate (package:serverpod_cli/src/generator/generator.dart:53:19)
<asynchronous suspension>
#6      StdOutLogger.progress (package:cli_tools/src/logger/loggers/std_out_logger.dart:141:20)
<asynchronous suspension>
#7      GenerateCommand.run (package:serverpod_cli/src/commands/generate.dart:67:17)
<asynchronous suspension>
#8      CommandRunner.runCommand (package:args/command_runner.dart:212:13)
<asynchronous suspension>
#9      BetterCommandRunner.runCommand (package:cli_tools/src/better_command_runner/better_command_runner.dart:140:7)
<asynchronous suspension>
#10     _main (file:///C:/Users/User/AppData/Local/Pub/Cache/hosted/pub.dev/serverpod_cli-2.0.0/bin/serverpod_cli.dart:59:3)   
<asynchronous suspension>
#11     main.<anonymous closure>
(file:///C:/Users/User/AppData/Local/Pub/Cache/hosted/pub.dev/serverpod_cli-2.0.0/bin/serverpod_cli.dart:33:9)
<asynchronous suspension>
#12     main (file:///C:/Users/User/AppData/Local/Pub/Cache/hosted/pub.dev/serverpod_cli-2.0.0/bin/serverpod_cli.dart:30:3)    
<asynchronous suspension>

PS C:\Users\User\vaani_server\vaani_server_server>

**To Reproduce**
1. serverpod generate

**Expected behavior**
After making changes in code, just called as routine, serverpod generate.
Expect to get complied source with no error as there was no syntax or other error(s) called out by Android Studio

**Screenshots**
![image](https://github.com/user-attachments/assets/e51eda45-5fc1-4f59-b3b6-9b8febb650f9)

**Serverpod versions**
serverpod 2.0.0

**Platform information**
MS Windows 10 Home Edition
ACER AMD laptop

**Additional context**
Made some change in one of the YAML files
Introduced a new endpoint for that YAML
Developed 2 methods in the endpoint file
Ran create-migration --force <-- no error
Ran serverpod generate --> Error
Tried twice, same result

**Comment by SandPod** [comment]
Hey @Sandyhi5!

It looks like we have run into some unexpected behavior when analyzing the endpoint you defined.

Can you share the code for the endpoints you created?
The endpoint method definition would be enough.

**Comment by Sandyhi5** [comment]
Right so here are the 2 methods:
class VnNotificationEndpoint extends Endpoint {

  // ...1
  Future<bool> createNotification( {
    required Session session,
    required int toNotifyId,
    required int byMemberId,
    int? articleId,
    int? responseId,
    String? partResp }
  ) async { ...return true; }

// ..2
  Future<List<VnNotification>> getMemberNotifications(Session session, int memberId) async { ...return }

--------------
Restarted Android Studio, then serverpod generate --> Error
Ran flutter pub get, then, serverpod generate  --> Error

**Comment by Sandyhi5** [comment]
I realized the error. Fixed it and it is working as expected now.
In the first method a change was required as below in session declaration:
Future createNotification(
Session session, {
required int toNotifyId,
required int byMemberId,
int? articleId,
int? responseId,
String? partResp }
) async { ...return true; }

It will be appreciated if the message was a little less foreboding. 

Thanks

**Comment by Sandyhi5** [comment]
Hi Gents

The ominous looking message is perhaps unnecessary. Will appreaciate a softer message.Problem is fixed now. It was a simple copy-paste from another .dart file
({required Session session, required int ... })-->(Session session, {required int ...})
I think the index error was for starting count with the first required field, which is normally not 'required' as it is a given
Will close Issue# 2920





Rgds
Sandeep Chandra
MD

(m) +61 414 398 198
(e):  ***@***.***
(w): www.hydrozen2050.com


Privileged/Confidential Information may be contained in this message and any documents attached to it. If you are not the intended recipient of this message, any use, disclosure or copying of this email is unauthorized. In such case, you should destroy this message, and notify us immediately. 

Any personal information in this email will be dealt within accordance wit the Privacy Act 1988 (Cth). Emails may be interfered with, may contain computer viruses or other defects and may not be successfully replicated on other systems. We give no warranties in relation to these matters and you receive this email and any attachments at your own risk. If you have any concerns, please contact us immediately.


 

    On Wednesday 30 October, 2024 at 07:42:33 pm GMT+11, Alexander Sandor ***@***.***> wrote:  
 
 


Hey @Sandyhi5!

It looks like we have run into some unexpected behavior when analyzing the endpoint you defined.

Can you share the code for the endpoints you created?
The endpoint method definition would be enough.

**Comment by Sandyhi5** [comment]
Closed

**Comment by SandPod** [comment]
If it is OK with you, @Sandyhi5, I would like to keep this issue open.

This is a bug in our analyzer that we should fix. As you stated, we should give a proper message stating what went wrong when this occurs.

### To summarize
The endpoint had the `Session` parameter declared as a named parameter instead of the first positional parameter of the method.

**Comment by Sandyhi5** [comment]
Totally OK. I can see it #2920 is now re-opened on GitHub
Also can I ask you about my next steps.
I intend to run my server on OCI [Oracle Cloud Infrastructure]]. What would you suggest I take care of while considering that. I know the default sql generated by serverpod is for PostGres SQL, but does that make it impossible / too hard / worth a try in your view to support on OCI?
Thanks




Rgds
Sandeep Chandra





Privileged/Confidential Information may be contained in this message and any documents attached to it. If you are not the intended recipient of this message, any use, disclosure or copying of this email is unauthorized. In such case, you should destroy this message, and notify us immediately. 

Any personal information in this email will be dealt within accordance wit the Privacy Act 1988 (Cth). Emails may be interfered with, may contain computer viruses or other defects and may not be successfully replicated on other systems. We give no warranties in relation to these matters and you receive this email and any attachments at your own risk. If you have any concerns, please contact us immediately.


 

    On Wednesday 30 October, 2024 at 09:30:08 pm GMT+11, Alexander Sandor ***@***.***> wrote:  
 
 


If it is OK with you, @Sandyhi5, I would like to keep this issue open.

This is a bug in our analyzer that we should fix. As you stated, we should give a proper message stating what went wrong when this occurs.

To summarize

The endpoint had the Session parameter declared as a named parameter instead of the first positional parameter of the method.

**Comment by SandPod** [comment]
@Sandyhi5 

> I know the default sql generated by serverpod is for PostGres SQL, but does that make it impossible / too hard / worth a try in your view to support on OCI

The framework currently only has a Postgress adapter meaning all queries generated by the framework are all adapted for Postgres. So, using the server together with a relational database that does not support the posgres flavour of SQL will not work. But It seems like OCI supports managed Postgres instances: https://www.oracle.com/cloud/postgresql/
